multy_core/src/ethereum/ethereum_stream.cpp

```diff
@@ -0,0 +1,120 @@
+/* Copyright 2018 by Multy.io
+ * Licensed under Multy.io license.
+ *
+ * See LICENSE for details
+ */
+
+#include "multy_core/src/ethereum/ethereum_stream.h"
+
+#include "multy_core/src/ethereum/ethereum_extra_data.h"
+
+#include "multy_core/src/exception_stream.h"
+#include "multy_core/src/utility.h"
+#include "multy_core/binary_data.h"
+
+#include <array>
+#include <regex>
+#include <unordered_map>
+#include <vector>
+#include <iostream>
+#include <iomanip>
+
+
+namespace
+{
+using namespace multy_core::internal;
+
+const size_t ETHEREUM_SIZE_VARIABLE_FUNCTION_CONTRACT = 32;
+static const std::array<const uint8_t, ETHEREUM_SIZE_VARIABLE_FUNCTION_CONTRACT> ZEROES = {0};
+
+} // namespace
+
+
+namespace multy_core
+{
+namespace internal
+{
+
+EthereumPayloadDataStream::EthereumPayloadDataStream()
+        :m_data()
+{
+    m_data.reserve(256);
+}
+
+EthereumPayloadDataStream::~EthereumPayloadDataStream()
+{
+}
+
+BinaryDataPtr EthereumPayloadDataStream::get_content()
+{
+    return make_clone(as_binary_data(m_data));
+}
+
+void EthereumPayloadDataStream::write_data(const void* data, size_t size)
+{
+    const uint8_t* d = reinterpret_cast<const uint8_t*>(data);
+    m_data.insert(m_data.end(), d, d + size);
+}
+
+EthereumPayloadAsUint256Stream::EthereumPayloadAsUint256Stream()
+        :m_data()
+{
+    m_data.reserve(256);
+}
+
+EthereumPayloadAsUint256Stream::~EthereumPayloadAsUint256Stream()
+{
+}
+
+BinaryDataPtr EthereumPayloadAsUint256Stream::get_content()
+{
+    return make_clone(as_binary_data(m_data));
+}
+
+void EthereumPayloadAsUint256Stream::write_data(const void* data, size_t size)
+{
+    INVARIANT(size <= ETHEREUM_SIZE_VARIABLE_FUNCTION_CONTRACT);
+    const uint8_t* d = reinterpret_cast<const uint8_t*>(data);
+    m_data.insert(m_data.end(), ZEROES.data(), ZEROES.data() + (ETHEREUM_SIZE_VARIABLE_FUNCTION_CONTRACT - size));
+    m_data.insert(m_data.end(), d, d + size);
+}
+
+EthereumPayloadAsUint256Stream& operator<<(EthereumPayloadAsUint256Stream& stream, const BinaryData& value)
+{
+    stream.write_data(value.data, value.len);
+    return stream;
+}
+
+EthereumPayloadAsUint256Stream& operator<<(EthereumPayloadAsUint256Stream& stream, const BigInt& value)
+{
+    const BinaryDataPtr data = value.export_as_binary_data(BigInt::EXPORT_BIG_ENDIAN);
+    if (data->len > ETHEREUM_SIZE_VARIABLE_FUNCTION_CONTRACT)
+    {
+        THROW_EXCEPTION2(ERROR_TRANSACTION_SPECIFIC_ERROR_BASE,
```

Comment:
Add an error code `ERROR_BIG_INT_TOO_BIG_FOR_UINT256` and use it here with message "BigInt value is too big for Ethereum smart contract method call."